File: app/src/main/java/jp/hazuki/yuzubrowser/webkit/CustomWebViewClient.java
package jp.hazuki.yuzubrowser.webkit;

import android.graphics.Bitmap;
import android.net.Uri;
import android.net.http.SslError;
import android.os.Message;
import android.view.KeyEvent;
import android.webkit.HttpAuthHandler;
import android.webkit.SslErrorHandler;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebView;
import android.webkit.WebViewClient;

public class CustomWebViewClient extends WebViewClient {
    public void doUpdateVisitedHistory(CustomWebView view, String url, boolean isReload) {
    }

    @Override
    public void doUpdateVisitedHistory(WebView view, String url, boolean isReload) {
        if (view instanceof CustomWebView)
            doUpdateVisitedHistory((CustomWebView) view, url, isReload);
    }

    public void onFormResubmission(CustomWebView view, Message dontResend, Message resend) {
        dontResend.sendToTarget();
    }

    @Override
    public void onFormResubmission(WebView view, Message dontResend, Message resend) {
        if (view instanceof CustomWebView)
            onFormResubmission((CustomWebView) view, dontResend, resend);
        else
            dontResend.sendToTarget();
    }

    public void onLoadResource(CustomWebView view, String url) {
    }

    @Override
    public void onLoadResource(WebView view, String url) {
        if (view instanceof CustomWebView)
            onLoadResource((CustomWebView) view, url);
    }

    public void onPageFinished(CustomWebView view, String url) {
    }

    @Override
    public void onPageFinished(WebView view, String url) {
        if (view instanceof CustomWebView)
            onPageFinished((CustomWebView) view, url);
    }

    public void onPageStarted(CustomWebView view, String url, Bitmap favicon) {
    }

    @Override
    public void onPageStarted(WebView view, String url, Bitmap favicon) {
        if (view instanceof CustomWebView)
            onPageStarted((CustomWebView) view, url, favicon);
    }

    public void onReceivedHttpAuthRequest(CustomWebView view, HttpAuthHandler handler, String host, String realm) {
        handler.cancel();
    }

    @Override
    public void onReceivedHttpAuthRequest(WebView view, HttpAuthHandler handler, String host, String realm) {
        if (view instanceof CustomWebView)
            onReceivedHttpAuthRequest((CustomWebView) view, handler, host, realm);
        else
            handler.cancel();
    }

    public void onReceivedLoginRequest(CustomWebView view, String realm, String account, String args) {
    }

    @Override
    public void onReceivedLoginRequest(WebView view, String realm, String account, String args) {
        if (view instanceof CustomWebView)
            onReceivedLoginRequest((CustomWebView) view, realm, account, args);
    }

    public void onReceivedSslError(CustomWebView view, SslErrorHandler handler, SslError error) {
        handler.cancel();
    }

    @Override
    public void onReceivedSslError(WebView view, SslErrorHandler handler, SslError error) {
        if (view instanceof CustomWebView)
            onReceivedSslError((CustomWebView) view, handler, error);
        else
            handler.cancel();
    }

    public void onScaleChanged(CustomWebView view, float oldScale, float newScale) {
    }

    @Override
    public void onScaleChanged(WebView view, float oldScale, float newScale) {
        if (view instanceof CustomWebView)
            onScaleChanged((CustomWebView) view, oldScale, newScale);
    }

    public void onUnhandledKeyEvent(CustomWebView view, KeyEvent event) {
    }

    @Override
    public void onUnhandledKeyEvent(WebView view, KeyEvent event) {
        if (view instanceof CustomWebView)
            onUnhandledKeyEvent((CustomWebView) view, event);
    }

    @Override
    public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
        if (view instanceof CustomWebView)
            return shouldInterceptRequest((CustomWebView) view, request);
        else
            return null;
    }

    public WebResourceResponse shouldInterceptRequest(CustomWebView view, WebResourceRequest request) {
        return null;
    }

    public boolean shouldOverrideKeyEvent(CustomWebView view, KeyEvent event) {
        return false;
    }

    @Override
    public boolean shouldOverrideKeyEvent(WebView view, KeyEvent event) {
        if (view instanceof CustomWebView)
            return shouldOverrideKeyEvent((CustomWebView) view, event);
        else
            return false;
    }

    public boolean shouldOverrideUrlLoading(CustomWebView view, String url, Uri uri) {
        return false;
    }

    @SuppressWarnings("deprecation")
    @Override
    public boolean shouldOverrideUrlLoading(WebView view, String url) {
        if (view instanceof CustomWebView)
            return shouldOverrideUrlLoading((CustomWebView) view, url, Uri.parse(url));
        else
            return false;
    }

    @Override
    public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
        if (view instanceof CustomWebView)
            return shouldOverrideUrlLoading((CustomWebView) view, request.getUrl().toString(), request.getUrl());
        else
            return false;
    }
}

File: module/webview/src/main/java/jp/hazuki/yuzubrowser/webview/CustomWebViewClient.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.webview

import android.annotation.TargetApi
import android.graphics.Bitmap
import android.net.Uri
import android.net.http.SslError
import android.os.Build
import android.os.Message
import android.view.KeyEvent
import android.webkit.*

open class CustomWebViewClient : WebViewClient() {
    open fun doUpdateVisitedHistory(web: CustomWebView, url: String, isReload: Boolean) {}

    override fun doUpdateVisitedHistory(view: WebView, url: String, isReload: Boolean) {
        if (view is CustomWebView)
            doUpdateVisitedHistory(view as CustomWebView, url, isReload)
    }

    open fun onFormResubmission(web: CustomWebView, dontResend: Message, resend: Message) {
        dontResend.sendToTarget()
    }

    override fun onFormResubmission(view: WebView, dontResend: Message, resend: Message) {
        if (view is CustomWebView)
            onFormResubmission(view as CustomWebView, dontResend, resend)
        else
            dontResend.sendToTarget()
    }

    open fun onLoadResource(web: CustomWebView, url: String) {}

    override fun onLoadResource(view: WebView, url: String) {
        if (view is CustomWebView)
            onLoadResource(view as CustomWebView, url)
    }

    open fun onPageFinished(web: CustomWebView, url: String) {}

    override fun onPageFinished(view: WebView, url: String) {
        if (view is CustomWebView)
            onPageFinished(view as CustomWebView, url)
    }

    open fun onPageStarted(web: CustomWebView, url: String, favicon: Bitmap?) {}

    override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
        if (view is CustomWebView)
            onPageStarted(view as CustomWebView, url, favicon)
    }

    open fun onPageChanged(web: CustomWebView, url: String, originalUrl: String, progress: Int, isLoading: Boolean) {}

    open fun onReceivedHttpAuthRequest(web: CustomWebView, handler: HttpAuthHandler, host: String, realm: String) {
        handler.cancel()
    }

    override fun onReceivedHttpAuthRequest(view: WebView, handler: HttpAuthHandler, host: String, realm: String) {
        if (view is CustomWebView)
            onReceivedHttpAuthRequest(view as CustomWebView, handler, host, realm)
        else
            handler.cancel()
    }

    open fun onReceivedLoginRequest(view: CustomWebView, realm: String, account: String?, args: String?) {}

    override fun onReceivedLoginRequest(view: WebView, realm: String, account: String?, args: String?) {
        if (view is CustomWebView)
            onReceivedLoginRequest(view as CustomWebView, realm, account, args)
    }

    open fun onReceivedSslError(web: CustomWebView, handler: SslErrorHandler, error: SslError) {
        handler.cancel()
    }

    override fun onReceivedSslError(view: WebView, handler: SslErrorHandler, error: SslError) {
        if (view is CustomWebView)
            onReceivedSslError(view as CustomWebView, handler, error)
        else
            handler.cancel()
    }

    open fun onReceivedError(view: CustomWebView, errorCode: Int, description: CharSequence, url: Uri) {}

    @TargetApi(Build.VERSION_CODES.M)
    override fun onReceivedError(view: WebView?, request: WebResourceRequest, error: WebResourceError) {
        if (request.isForMainFrame && view is CustomWebView)
            onReceivedError(view as CustomWebView, error.errorCode, error.description, request.url)
    }

    @Suppress("OverridingDeprecatedMember")
    override fun onReceivedError(view: WebView?, errorCode: Int, description: String, failingUrl: String) {
        if (view is CustomWebView)
            onReceivedError(view as CustomWebView, errorCode, description, Uri.parse(failingUrl))
    }

    open fun onScaleChanged(view: CustomWebView, oldScale: Float, newScale: Float) {}

    override fun onScaleChanged(view: WebView, oldScale: Float, newScale: Float) {
        if (view is CustomWebView)
            onScaleChanged(view as CustomWebView, oldScale, newScale)
    }

    open fun onUnhandledKeyEvent(view: CustomWebView, event: KeyEvent) {}

    override fun onUnhandledKeyEvent(view: WebView, event: KeyEvent) {
        if (view is CustomWebView)
            onUnhandledKeyEvent(view as CustomWebView, event)
    }

    override fun shouldInterceptRequest(view: WebView, request: WebResourceRequest): WebResourceResponse? {
        return if (view is CustomWebView)
            shouldInterceptRequest(view as CustomWebView, request)
        else
            null
    }

    open fun shouldInterceptRequest(web: CustomWebView, request: WebResourceRequest): WebResourceResponse? = null

    open fun shouldOverrideKeyEvent(web: CustomWebView, event: KeyEvent): Boolean = false

    override fun shouldOverrideKeyEvent(view: WebView, event: KeyEvent): Boolean {
        return if (view is CustomWebView)
            shouldOverrideKeyEvent(view as CustomWebView, event)
        else
            false
    }

    open fun shouldOverrideUrlLoading(web: CustomWebView, url: String, uri: Uri): Boolean = false

    @Suppress("OverridingDeprecatedMember")
    override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
        return if (view is CustomWebView)
            shouldOverrideUrlLoading(view as CustomWebView, url, Uri.parse(url))
        else
            false
    }

    override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
        return if (view is CustomWebView)
            shouldOverrideUrlLoading(view as CustomWebView, request.url.toString(), request.url)
        else
            false
    }

    open fun onPageCommitVisible(web: CustomWebView, url: String) {}

    override fun onPageCommitVisible(view: WebView?, url: String?) {
        if (view is CustomWebView && url != null) {
            onPageCommitVisible(view as CustomWebView, url)
        }
    }
}
